src/Form/LinkBlockFormDataProvider.php

```diff
@@ -0,0 +1,283 @@
+     * @param array $languages
+     * @param int $shopId
+     */
+    public function __construct(
+        LinkBlockRepository $repository,
+        LinkBlockCacheInterface $cache,
+        ModuleRepository $moduleRepository,
+        array $languages,
+        $shopId
+    ) {
+        $this->repository = $repository;
+        $this->cache = $cache;
+        $this->moduleRepository = $moduleRepository;
+        $this->languages = $languages;
+        $this->shopId = $shopId;
+    }
+
+    /**
+     * @return array
+     *
+     * @throws \PrestaShopDatabaseException
+     * @throws \PrestaShopException
+     */
+    public function getData()
+    {
+        if (null === $this->idLinkBlock) {
+            return [];
+        }
+
+        $linkBlock = new LinkBlock($this->idLinkBlock);
+
+        $arrayLinkBlock = (array) $linkBlock;
+
+        //The form and the database model don't have the same data hierarchy
+        //Transform array $custom[en][1][name] to $custom[1][en][name]
+        $arrayCustom = [];
+        foreach ($arrayLinkBlock['custom_content'] as $idLang => $customs) {
+            if (empty($customs)) {
+                continue;
+            }
+
+            foreach ($customs as $i => $custom) {
+                $arrayCustom[$i][$idLang] = $custom;
+            }
+        }
+
+        return ['link_block' => [
+            'id_link_block' => $arrayLinkBlock['id_link_block'],
+            'block_name' => $arrayLinkBlock['name'],
+            'id_hook' => $arrayLinkBlock['id_hook'],
+            'cms' => $arrayLinkBlock['content']['cms'],
+            'product' => $arrayLinkBlock['content']['product'],
+            'static' => $arrayLinkBlock['content']['static'],
+            'custom' => $arrayCustom,
+        ]];
+    }
+
+    /**
+     * @param array $data
+     *
+     * @return array
+     *
+     * @throws \PrestaShop\PrestaShop\Adapter\Entity\PrestaShopDatabaseException
+     */
+    public function setData(array $data)
+    {
+        $linkBlock = $data['link_block'];
+        $errors = $this->validateLinkBlock($linkBlock);
+        if (!empty($errors)) {
+            return $errors;
+        }
+        $customContent = [];
+        if (!empty($linkBlock['custom'])) {
+            foreach ($linkBlock['custom'] as $customIndex => $customLanguages) {
+                foreach ($customLanguages as $idLang => $custom) {
+                    $customContent[$idLang][$customIndex] = $custom;
+                }
+            }
+        }
+
+        if (empty($linkBlock['id_link_block'])) {
+            $linkBlockId = $this->repository->create(
+                $linkBlock['block_name'],
+                $linkBlock['id_hook'],
+                $linkBlock['cms'],
+                $linkBlock['static'],
+                $linkBlock['product'],
+                $customContent
+            );
+        } else {
+            $linkBlockId = $linkBlock['id_link_block'];
+            $this->repository->update(
```

Comment:
why passing all parameters for create and update, could be
`update($id, $data)` and `create($data)` wdyt?